### PR TITLE
Pull request for lesson 4 exercise 2 "the mailroom part 2"

### PR DESCRIPTION
**Pull request by nskvarch**
Pull request for lesson 4 exercise 2 "the mailroom part 2" and lesson 4 exercise 3 "Trigrams : Kata Fourteen"
Also contains the files for exercise 1 "the dict lab" and supporting files for the trigrams exercise

Based on instructor feedback, corrected and fixed:
in Trigrams:
created a random seed rather than a fixed one, the story is now completely different on each run
in Mairoom2.py
Added the list option to the add a contribution menu 
added the correct sorting feature, not on name but on total donated in descending order
fixed the database handling so the total amount donated is correctly displayed in the donor reports

Update 12_10 added files for the lesson 5 exercise "Exceptions"

Update 12_10 added the files for lesson 5 exercise 4 "the mailroom part 3" exceptions and modules.

### CODE REVIEW

File: students/nskvarch/lesson4/mailroom2.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+def run_report(donor_db):
+    """take in the donor database, create organize and sort a table in the form of a list and then hand the table off
+    to the printer function """
+    data_lines = []
+    for name in donor_db:
+        data = donor_db[name]
+        data_lines.append((name, str(max(data)), str(len(data)), str(sum(data) / len(data))))
+
+    to_print = [("Name", "Total Donated", "Times Donated", "Average Donation")] + sorted(data_lines, key=lambda x: x[0])
+    print_report(to_print)
```
Comment (charlesrobison) [inline students/nskvarch/lesson4/mailroom2.py]: This should be sorting the donors by total donated, in descending order.

File: students/nskvarch/lesson4/mailroom2.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+
+def thanks(donor_db):
+    """create a thank you note customized to the name provided, 
+    if the name was not in the donor database already, add the name and promt for a donation ammount"""
+    name = input("\nPlease enter the name of the person donating: ")
```
Comment (charlesrobison) [inline students/nskvarch/lesson4/mailroom2.py]: This should also give the user the option to get a list of current donors in the database.

File: students/nskvarch/lesson4/mailroom2.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    ammount = float(input("\nPlease enter the ammount donated:  "))
+    for donor_name in donor_db:
+        if name.lower() == donor_name.lower():
+            donor_db[donor_name].append(ammount)
```
Comment (charlesrobison) [inline students/nskvarch/lesson4/mailroom2.py]: This is overriding the total at the moment, not adding to it.

File: students/nskvarch/lesson4/trigrams.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        sys.exit(1)
+
+    in_data = read_in_data(filename)
+    words = make_words(in_data)
+    trigrams, seed = build_trigram(words)
+    new_text = build_text(trigrams, seed)
+    text_string = write_text(new_text)
+    
+
+    print(text_string)
```
Comment (charlesrobison) [inline students/nskvarch/lesson4/trigrams.py]: This doesn't seem to be functioning as expected.  Running this should create a new story from the existing text, but this is currently recreating the same text in the same order.

File: students/nskvarch/lesson4/mailroom2.py (hunk 1 above)
Comment (nskvarch) [reply students/nskvarch/lesson4/mailroom2.py]: Corrected the sort to sort by the total donated in descending order

File: students/nskvarch/lesson4/mailroom2.py (hunk 2 above)
Comment (nskvarch) [reply students/nskvarch/lesson4/mailroom2.py]: added the list option to the sub menu for this item

File: students/nskvarch/lesson4/mailroom2.py (hunk 3 above)
Comment (nskvarch) [reply students/nskvarch/lesson4/mailroom2.py]: found I was handing the max value rather then the sum value to the sort operation

File: students/nskvarch/lesson4/trigrams.py (hunk 4 above)
Comment (nskvarch) [reply students/nskvarch/lesson4/trigrams.py]: It was always using the same first six or seven words then the rest of the story was random, changed my seed for the word list builder to be random and now it builds a completely random story each time.

**Comment by natasha-aleksandrova** [comment]
@charlesrobison looks like there are some new commits - can you please take another look?

**Comment by nskvarch** [comment]
Added files for the lesson 5 exercise "exceptions"

**Comment by natasha-aleksandrova** [comment]
@charlesrobison please do another review :)

**Pull request merged**